### PR TITLE
IGNITE-21238 Add backfill index status to catalog

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-21238

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java [hunk 1]
```diff
@@ -101,8 +105,8 @@
  *
  * <p>Approximate recovery algorithm:</p>
  * <ul>
- *     <li>For registered indexes: <ul>
- *         <li>If the new index did not have time to add
+ *     <li>For buildings indexes: <ul>
```
Comment (rpuch) [inline modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java:108]:
```suggestion
 *     <li>For building indexes: <ul>
```

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildingManager.java [hunk 2]
```diff
@@ -34,12 +48,18 @@
 import org.apache.ignite.network.ClusterService;
 
 /**
- * Component is responsible for building indexes and making them {@link CatalogIndexDescriptor#available() available}. Both in a running
+ * Component is responsible for building indexes and making them in {@link CatalogIndexStatus#AVAILABLE available}. Both in a running
```
Comment (rpuch) [inline modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildingManager.java:51]:
```suggestion
 * Component is responsible for building indexes and making them {@link CatalogIndexStatus#AVAILABLE available}. Both in a running
```

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexChooser.java [hunk 3]
```diff
@@ -158,7 +160,7 @@ public void close() {
      */
     List<CatalogIndexDescriptor> chooseForRwTxUpdateOperation(int catalogVersion, int tableId) {
         return inBusyLock(busyLock, () -> {
-            List<CatalogIndexDescriptor> tableIndexes = catalogService.indexes(catalogVersion, tableId);
+            List<CatalogIndexDescriptor> tableIndexes = collectAvailableAndBuildingIndexes(catalogService.indexes(catalogVersion, tableId));
```
Comment (rpuch) [inline modules/index/src/main/java/org/apache/ignite/internal/index/IndexChooser.java:163]: We must write to indices that were REGISTERED at the tx start as well

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogIndexStatus.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.catalog.descriptors;
+
+/**
+ * Index statuses.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogIndexStatus.java:21]: fix it

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java (hunk 1 above)
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java:108]: fix it

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildingManager.java (hunk 2 above)
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildingManager.java:51]: fix it

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/ChangeIndexStatusValidationException.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.catalog;
+
+import org.apache.ignite.internal.catalog.descriptors.CatalogIndexDescriptor;
+import org.apache.ignite.internal.catalog.descriptors.CatalogIndexStatus;
+
+/** Exception that occurs when an incorrect attempt to change the {@link CatalogIndexDescriptor#status()}. */
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/ChangeIndexStatusValidationException.java:23]: fix it

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogIndexStatus.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ * <ul>
+ *     <li>{@link #REGISTERED} -> {@link #BUILDING} -> {@link #AVAILABLE}.</li>
+ *     <li>{@link #AVAILABLE} (PK index).</li>
+ * </ul>
+ */
+public enum CatalogIndexStatus {
+    /**
+     * Index has been registered and is awaiting the start of building.
+     *
+     * <p>Not readable or writable.</p>
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogIndexStatus.java:33]: Fix it

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexChooser.java (hunk 3 above)
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexChooser.java]: fix it

**Review by rpuch — APPROVED** [review]

**Pull request merged**